tensor: Implementation of tensorial and symbolic partial derivatives

joha2 opened the pull request:
~Implemented `_eval_derivatives` for `TensAdd`, `TensMul`, and `Tensor`.~
~Further, added `_diff_wrt = True` and `is_scalar = True` to `Tensor`.~

The implementation for `_eval_partial_derivative` and `_expand_partial_derivative` 
are more or less taken from
`Mul` and `Add`. Derivatives with respect to tensors are implemented
in such a manner, that a covariant index in the derivative is
counted contravariant, and vice versa. Therefore, `PartialDerivative(a(i), a(i))._perform_derivative()`
is a perfectly valid expression.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #18048 

#### Brief description of what is fixed or changed

Definitions:
```
>>> Euclidean = TensorIndexType('Euclidean', dummy_name='E')
>>> a = tensor_heads('a', [Euclidean])
>>> b = tensor_heads('b', [Euclidean])
>>> i, j, k = tensor_indices('i j k', Euclidean)
```
Mixed tensor (scalar) and symbol expression, calculating derivative with respect to symbol:
```
>>> expr2 = a(i)*a(-i) + tau**2 + tau*b(i)*b(-i)
>>> PartialDerivative(expr2, tau)._perform_derivative() # current PR
2*tau + b(E_0)*b(-E_0)
>>> expr2.diff(tau).doit() # current master
Derivative(a(E_0)*a(-E_0) + tau**2 + tau*b(E_0)*b(-E_0), tau)
```
Expression gets not evaluated in current master.

Mixed tensor (scalar) and symbol expression, calculating derivative with respect to tensor:
```
>>> expr3 = PartialDerivative(expr2, a(k))
>>> expr3._perform_derivative().contract_metric(i.tensor_index_type.metric).canon_bp() # current PR
2*KD(-k, E_0)*a(-E_0)
>>> expr2.diff(a(k)).doit().contract_metric(i.tensor_index_type.metric).canon_bp() # current master
Traceback (most recent call last):
  File "sympy_issue_tensor_report.py", line 89, in <module>
    diffmyexpr3 = myexpr2.diff(a(k))
  File "/home/joha2/Programming/git/sympy/sympy/core/expr.py", line 3352, in diff
    return Derivative(self, *symbols, **assumptions)
  File "/home/joha2/Programming/git/sympy/sympy/core/function.py", line 1343, in __new__
    __)))
ValueError: 
Can't calculate derivative wrt a(k).
```
Leads to error in current master, since `_diff_wrt = False`.

#### Other comments

It is assumed that free indices (i.e. `get_free_indices()` of `Tensor`) are always
in the correct order such that for `diff(a(i, j), a(k, l))` both args[1]
give `(i, j)` and `(k, l)`, respectively.

Further, the implementation for tensorial derivates is -
besides fulfilling product and sum rule - given by (**):

* \partial a_i/\partial a_k = delta_i^k
* \partial a_i/\partial a^k = g_{ij} delta^j_k
* \partial a^i/\partial a^k = delta^i_k
* \partial a^i/\partial a_k = g^{ij} delta_j^k

##### Known issues and open problems

* ~Does `is_scalar = True` lead to any inconsistencies?~ -- removed
* What about anticommuting/non-commuting tensors?
  Is there anything special to be considered?
* (**) What if there exists no metric? `diff(a(i), a(-k)) = 0`?
  Is this case even possible/considered in `tensor`?
* ~`sqrt(a(i)*a(-i))` cannot be derived by `a(i)` since `__div__`
  allows no tensor expressions, see (***).~ -- later

##### Examples

Definitions:
```
>>> Euclidean = TensorIndexType('Euclidean', dummy_name='E')
>>> a = tensor_heads('a', [Euclidean])
>>> b = tensor_heads('b', [Euclidean])
>>> i, j, k = tensor_indices('i j k', Euclidean)
```
Derivative with respect to symbol as coefficient:
```
>>> expr1 = a(i) + tau**2*b(i)
>>> PartialDerivative(expr1, tau)._perform_derivative()
2*tau*b(i)
```
Derivative of mixed symbol and scalar tensor expression with respect to symbol:
```
>>> expr2 = a(i)*a(-i) + tau**2 + tau*b(i)*b(-i)
>>> PartialDerivative(expr2, tau)._perform_derivative()
2*tau + b(E_0)*b(-E_0)
```
Derivative of mixed symbol and scalar tensor expression with respect to tensor:
```
>>> expr3 = PartialDerivative(expr2, a(k))
>>> expr3._perform_derivative().contract_metric(i.tensor_index_type.metric).canon_bp()
2*KD(-k, E_0)*a(-E_0)
```
Verifying that divergence of vector with its own components leads to number of dimenions.
```
>>> expr4 = a(i)
>>> PartialDerivative(expr4, a(i))._perform_derivative().contract_metric(i.tensor_index_type.metric).\
    contract_delta(i.tensor_index_type.delta)
dim_E
```

##### TODO

* doctests to be implemented
* ~tests to be implemented~

##### Future nice to have features

* Implement an abstract partial/covariant tensor derivative, e.g.:
  \partial_i (A_k*B_l) = (\partial_i A_k) B_l + A_k (\partial_i B_l)
  (Like `PD[][]` and `CD[][]` in `xTensor` [1])
* (***) Improve tensorial scalar functions support,
  see `xTensor` [1] `Scalar[]` head.

##### References

[1] http://xact.es/xTensor

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
     * Implemented derivatives with respect to symbols in tensor expressions
     * Implemented derivatives with respect to tensors in tensor expressions
<!-- END RELEASE NOTES -->

joha2 committed:
tensor: Implemented tensorial and symbolic derivatives

Implemented `_eval_derivatives` for `TensAdd`, `TensMul`, and `Tensor`.
Further, added `_diff_wrt = True` and `is_scalar = True` to `Tensor`.

The implementation for `_eval_derivatives` are more or less taken from
`Mul` and `Add`. Derivatives with respect to tensors are implemented
in such a manner, that a covariant index in the derivative is
counted contravariant, and vice versa. Therefore, `diff(a(i), a(i))`
is a perfectly valid expression.

It is assumed that free indices (i.e. args[1] of `Tensor`) are always
in the correct order such that for `diff(a(i, j), a(k, l))` both args[1]
give (i, j) and (k, l), respectively.

Further, the implementation for tensorial derivates is -
besides fulfilling product and sum rule - given by (**):

* \partial a_i/\partial a_k = delta_i^k
* \partial a_i/\partial a^k = g_{ij} delta^j_k
* \partial a^i/\partial a^k = delta^i_k
* \partial a^i/\partial a_k = g^{ij} delta_j^k

Known issues and open problems:

* Does `is_scalar = True` lead to any inconsistencies?
* What about anticommuting/non-commuting tensors?
  Is there anything special to be considered?
* (**) What if there exists no metric? `diff(a(i), a(-k)) = 0`?
  Is this case even possible/considered in `tensor`?
* `sqrt(a(i)*a(-i))` cannot be derived by `a(i)` since `__div__`
  allows no tensor expressions, see (***).

Examples:

>>> Euclidean = TensorIndexType('Euclidean', dummy_name='E')
>>> a = tensor_heads('a', [Euclidean])
>>> b = tensor_heads('b', [Euclidean])
>>> i, j, k = tensor_indices('i j k', Euclidean)
>>> expr1 = a(i) + tau**2*b(i)
>>> expr1.diff(tau).doit()
2*tau*b(i)
>>> expr2 = a(i)*a(-i) + tau**2 + tau*b(i)*b(-i)
>>> expr2.diff(tau).doit()
2*tau + b(E_0)*b(-E_0)
>>> expr3 = myexpr2.diff(a(k))
>>> expr3.doit().contract_metric(i.tensor_index_type.metric).canon_bp()
2*KD(-k, E_0)*a(-E_0)
>>> expr4 = a(i)
>>> expr4.diff(a(i)).contract_metric(i.tensor_index_type.metric).\
    contract_delta(i.tensor_index_type.delta)
dim_E

TODO:

* doctests to be implemented
* tests to be implemented

Future nice to have features:

* Implement an abstract partial/covariant tensor derivative, e.g.:
  \partial_i (A_k*B_l) = (\partial_i A_k) B_l + A_k (\partial_i B_l)
  (Like `PD[][]` and `CD[][]` in `xTensor` [1])
* (***) Improve tensorial scalar functions support,
  see `xTensor` [1] `Scalar[]` head.

References:

[1] http://xact.es/xTensor

Upabjojr commented:
Hi @joha2 ,

Thanks for your contribution.

I would suggest to handle the tensor module in a special way as concerns derivatives. In particular, we want to make the module `sympy.tensor.tensor` independent of the coordinate system representation.

Partial derivatives are non-covariant (i.e. when you take the derivative in one coordinate system you may end having a different expression if you take the same derivative in another coordinate system). In particular, when you contract all indices, the resulting quantity has to be independent of the coordinate system. In tensor calculus ___covariant derivatives___ are introduced for this purpose.

What you are doing is the computation of ___partial derivatives___.

My suggestion:
- `_eval_derivative` should return an error explaining the problem with the tensor module.
- your implementations of `_eval_derivative` should be renamed to something else, e.g. `_eval_partial_derivative`.
- A new method called `.diff_partial( )` should be added, which calls `._eval_partial_derivative`.

I'm actually open to discuss this issue if there are better ideas. I'm open to other API designs, I just want to avoid making ___partial derivatives___ the default in this module.

joha2 commented:
Hey @Upabjojr Thanks for review! As I mentioned in #18048 there are two points to be discussed. If there at the end is no compromise possible, we could still rely on the diff_partial method:

* If a tensor expression is derived by a symbol, the character of the expression is not altered and therefore it is not a partial derivative introducing a coordinate dependency, right? (E.g. deriving by the proper time parameter in general relativity)
* If a tensor expression is derived by a tensor quantity (not coordinates of a specific basis), this should also be a valid expression, since I could define this derivative by performing a first order expansion in a scalar symbol, right?
* If a tensor expression is derived by coordinates and the tensors are considered as fields/densities, then we have the problem, you described. 

In short I understood it in the following: `diff(a(i), a(j))` and `diff(a(i), tau)` should be no problem due to the first point, but `diff(a(i), x(j))` where `x` is a local coordinate, is problematic. So maybe, it makes sense to distuingish between "coordinate" tensors and "tensorfield/densities" by introducing additional flags/parameters into `Tensor`? Maybe also a `dependencies` list in `Tensor` could help.

Best regards
Johannes

Upabjojr commented:
- `diff(a(i), a(j))` may behave quite differently: think of Cartesian and spherical coordinates. I'm not sure that a coordinate transformation preserves this quantity under contractions (i.e. trace).
- `diff(a(i), tau)`, actually, [first result](https://en.wikiversity.org/wiki/Operator_of_proper-time-derivative) I got, it looks like the proper time derivative is expressed in terms of covariant derivatives in general relativity.

> therefore it is not a partial derivative introducing a coordinate dependency, right? (E.g. deriving by the proper time parameter in general relativity)

Take again the spherical coordinates as an example: you cannot take the norm of `position.diff(x(i))` as the velocity anymore (you can in Cartesian coordinates).

> If a tensor expression is derived by a tensor quantity (not coordinates of a specific basis), this should also be a valid expression, since I could define this derivative by performing a first order expansion in a scalar symbol, right?

> If a tensor expression is derived by coordinates and the tensors are considered as fields/densities, then we have the problem, you described.

The expression `.diff(x(i))` is just syntactic sugar for derivatives by the coordinates. Indeed, `x(i)` is meant to represent the vector of coordinates. With `x(i)` you mean `x`, `y`, `z`, usually.

joha2 commented:
Your concerns are perfectly valid, if you consider coordinates as tensor components. Although, I wanted to perform derivatives if I explicit consider no coordinates just the tensor fields themselves or some scalar parameter. I have to go now. But I will give a more thorough answer later. At the end of the day, it is maybe a good idea to just relabel to `diff_partial`, but maybe it is not immediately necessary.

Best wishes
Johannes

Upabjojr commented:
> At the end of the day, it is maybe a good idea to just relabel to diff_partial, but maybe it is not immediately necessary.

I had some thoughts time ago, and I was thinking of implementing something like:

```python

class TensPartialDerivative(TensExpr):

    def __init__(self, expr, x):
        ...

    def doit(self, *args, **kwargs):
        if isinstance(expr, TensMul):
            ...
        elif isinstance(expr, TensAdd):
            ...
        ...
```

This way you clearly state what you are doing. The advantage of having a class instead of a function is to allow the unevaluated form: sometimes you want to have derivatives in your formulas without computing their values.

joha2 committed:
tensor: Introduced _eval_partial_derivative. Added class TensPartialDerivative

This commit takes the following points of discussion into account:

* `diff` should lead to coordinate invariant expressions
* this cannot be guaranteed for partial derivatives
* `_eval_derivative` should be renamed into `_eval_partial_derivative`
* a new class `TensPartialDerivative` should be introduced
* this could provide a method for handling abstract partial derivatives

Therefore, the method `_eval_derivative` was renamed to
`_eval_partial_derivative`. This lead to some new `if` statements
(for taking care, whether a `TensExpr` is differentiated or some other
entity), since one could not rely on the `diff` functionality of the super
classes very well.

Further the `TensPartialDerivative` class (a subclass of `TensExpr`) is
introduced, where the `doit` method calls `_eval_partial_derivative`.
The indexing of the unperformed abstract derivative remains to be discussed.

Upabjojr commented on sympy/tensor/tensor.py:
```diff
@@ -2538,9 +2538,14 @@ def __iter__(self):
+    def _eval_partial_derivative(self, s):
         # Evaluation like Add
-        return self.func(*[a.diff(s) for a in self.args])
+        # TODO: what about if s is a tensor and a no tensor expression?
```
This should be handled in `TensPartialDerivative`, I wouldn't care about handling types in a hidden method.

Upabjojr commented on sympy/tensor/tensor.py:
```diff
@@ -2919,44 +2924,48 @@ def _eval_rewrite_as_Indexed(self, tens, indices):
-            return TensMul.fromiter(mykroneckerdeltas)
-        else:
+    def _eval_partial_derivative(self, s):
+        if not isinstance(s, Tensor):
```
This can be handled in `TensPartialDerivative`.

Upabjojr commented on sympy/tensor/tensor.py:
```diff
@@ -3802,6 +3818,38 @@ def _extract_data(self, replacement_dict):
+        #elif isinstance(expr, Tensor):
+        #    ... code for TensAdd expr being derived by v ...
+        #    return result
+        return expr._eval_partial_derivative(v)
```
OK!

Upabjojr commented:
Actually... `PartialDerivative` is already implemented:
https://github.com/sympy/sympy/blob/master/sympy/tensor/toperators.py#L4

joha2 committed:
Merge branch upstream master into 18048_tensor_derivatives

Due to a fix in PartialDerivative which has to be incorporated,
the upstream master had to be merged into this feature branch.

joha2 committed:
Added doit() to PartialDerivative, removed TensPartialDerivative

joha2 commented:
@Upabjojr implementing the `doit()` method leads to difficulties in one of the tests:
```Python
    expr2b = A(i)*PartialDerivative(H(k, -i), A(-j))
    assert expr2b.get_indices() == [L_0, k, -L_0, j]
```

Since writing a `*` instead of `TensMul` seems to perform a `doit()` on both expressions, this evaluates to zero and leads to a failing test, since a `0` has no index structure. I would suggest to let doit() only expand products or sums of partial derivatives without evaluating them and let the evaluation done by another method of PartialDerivative. A dirty work around without touching the doit would be to use `TensMul` directly, not `*`. What do you think?

Upabjojr commented:
Don't write the code into .doit(), just create a separate method.

joha2 committed:
tensor: PartialDerivative: Restored doit(), moved code into new function perform()

joha2 commented:
TODO:
* abstract derivative of product and sum rule in doit()
* derivatives of symbols (condition in index check code)

joha2 committed:
tensor: PartialDerivative: implemented symbol and tensor derivatives, even in a mixed fashion

joha2 committed:
tensor: PartialDerivative: doit() implemented to perform sums, products, multiple derivatives

TODO:
* perform() needs testing
* doit() needs testing

joha2 committed:
Merge branch 'master' of https://github.com/sympy/sympy into 18048_tensor_derivatives

joha2 commented:
@Upabjojr I've commited some changes. Could you please review them?
* perform() now performs the derivative, where only explicit dependencies are taken into account
* doit() performs an abstract derivative, where now derivative vanishes
* both can perform multi derivatives

Anyway, I had some strange error, due to a non-existing `_index_structure` of `TensAdd` and `PartialDerivative`, which I am not sure how to fix:
```Python
>>> (PartialDerivative(p(mu)*q(nu), tau) - PartialDerivative(p(mu), tau)*q(nu) - PartialDerivative(q(nu), tau)*p(mu)).doit()
<snip>
<snap>
AttributeError: 'TensAdd' object has no attribute '_index_structure'
```
and indeed:
```Python
>>> print((q(mu)*p(nu))._index_structure)
_IndexStructure([(mu, 0), (nu, 1)], [], [Euclidean, Euclidean])
>>> print((q(mu)+p(mu))._index_structure)
AttributeError: 'TensAdd' object has no attribute '_index_structure'
>>> print(PartialDerivative(p(mu), q(nu))._index_structure)
AttributeError: 'PartialDerivative' object has no attribute '_index_structure'
```
Should not all tensor constructions return an `_index_structure`, shouldn't they? In particular due to the function `get_indices()` which asks explicitly for an `TensExpr` instance, which means that all derived types have to have this attribute. I'am just to inexperienced with the tensor API to understand how it is supposed to work, to fix those things by myself.

Upabjojr commented:
TensAdd does not have sorted indices. Just think of the case A_ij + B_ji... there is no index order. The variable _index_structure was meant to solve this issue.

joha2 commented:
@Upabjojr so does this mean, `_index_structure` is missing in `TensAdd`, or does it mean `TensAdd` needs no `_index_structure`?

Btw: The tests fail due to `PartialDerivative.doit()` which somehome arranges the indices after performing the actual derivatives. So e.g. for `expr3` in your tests, the derivative index is not at the last position anymore. Should we compare to some sorted list, to exclude such failures? Or should the `doit()` functionality (which means expansion of the derivatives) be put into another function?
A third possibility would be to use `TensMul(A(i), PartialDerivative(...))` instead of `A(i)*PartialDerivative(...)`, but I don't know if this is intended.

Upabjojr commented on sympy/tensor/toperators.py:
```diff
@@ -56,27 +56,75 @@ def __new__(cls, expr, *variables):
         return args, indices, free, dum
 
     def doit(self):
+        """
```
The documentation for `.doit()` should be inherited from SymPy's core.

joha2 replied:
Ok, I remove my own comment.

joha2 committed:
tensor: PartialDerivative: removed doc string from doit()

joha2 committed:
tensor: PartialDerivative: fixed tests

For expr3a the execution of doit() is prevented by using TensMul,
afterwards the index structure is compared to the expected value.
In expr3b doit() is explicitly performed and it is checked whether
the derivative index is still at the right position and the free
vector index, too.

Upabjojr commented:
Concerning the problem with `_index_structure`:

https://github.com/sympy/sympy/pull/18224

Upabjojr commented on sympy/tensor/tensor.py:
```diff
@@ -2918,6 +2930,49 @@ def _eval_rewrite_as_Indexed(self, tens, indices):
+            # @a^i/@a_k = g^ij delta_j^k
+            # TODO: if there is no metric present, the derivative should be zero?
+
+            if selfhead is otherhead:
```
never use `is`, use rather `==`

joha2 replied:
Ok, will fix

Upabjojr commented:
@joha2 we need a lot of unit tests.

My usual standard is to write:
```python
expr = ... some expression ...
assert expr.some_operation() == expected result
```

I would suggest to start writing the unit tests, so we have a reference of what result we want to have.

For example:
```python
expr = PartialDerivative(A(i), A(j))
assert expr.doit() == KroneckerDelta(i, j)

expr = PartialDerivative(A(i), B(j))
assert expr.doit() == expr  # don't evaluate if don't know how
```

Upabjojr commented:
I think `.doit()` should compute the deltas. `.expand()` is probably the method that should not compute the deltas.

joha2 commented:
@Upabjojr OK, this means, I will wait until your PR is accepted an integrate it into my branch, like last time.

Thanks for the clarification, on how the unit tests are meant.

Let me shortly summarize, whether I understood correctly about the naming conventions of the functions:
* `doit()` should calculate the derivatives and lead to Kronecker deltas if tensor is derived by tensor (this means, it corresponds to `_eval_partial_derivative`)
* `expand()` should expand the partial derivatives like it is now done in `doit()` (i.e. sum rule, product rule, ...)

This means one has to be careful to write down test, where `doit()` is automatically performed like for `A(i)*PartialDerivative(C(k), D(j))`, right?

Upabjojr commented:
@joha2 I'm actually not sure what's the best way to implement this, it's a subjective issue.

I would suggest to put everything in a private method, for now. For example, `._evaluate_partial_derivative()`. When we decide if it has to be called in `.doit()`, we will simply put a link.

joha2 commented:
Yeah, it is subjective. But depending on what is in the `doit()` method, we have to craft the tests. I also would have no problem, if the `doit()` method is never implemented, and expansion as well as `_eval_partial_derivative` are called by other names. Anyway, the user should have the possibility to easily perform these tasks.

joha2 committed:
tensor: PartialDerivative: fixed pretty print

Further substituted `is` by `==` in `Tensor._eval_partial_derivative`.

joha2 commented:
@Upabjojr do you have an idea, why here the tensor_flow test is crashing? I did not do anything with the code.

oscarbenjamin commented:
> @Upabjojr do you have an idea, why here the tensor_flow test is crashing? I did not do anything with the code.

That is #18090. I restarted the job and it will probably pass second time round.

joha2 commented:
@oscarbenjamin thanks!

joha2 commented:
@Upabjojr so, what do you think: should we move the `doit()` functionality into some function `expand()`? Such that both expansion and performance of a derivative can be called by the user?

Upabjojr commented:
I would suggest some private methods for now.

joha2 commented:
ok

Upabjojr commented:
This other PR is merged.

joha2 committed:
Merge branch 'master' of https://github.com/joha2/sympy into 18048_tensor_derivatives

joha2 committed:
tensor: PartialDerivative: doit() does nothing, added _eval_partial_derivative() and _expand_partial_derivative()

_eval_partial_derivative(): performs the derivatives of a tensor expression explicitly.
_expand_partial_derivative(): implements product and sum rule for tensor expressions.

joha2 commented:
@Upabjojr verification of product rule still not works. Within the sum TensAdd complains about not having the attribute `components`. With the new commit it looks like:
```Python
(PartialDerivative(p(mu)*q(nu), tau)._expand_partial_derivative() - PartialDerivative(p(mu), tau)*q(nu) - PartialDerivative(q(nu), tau)*p(mu))

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-15-a0482c523c1b> in <module>()
----> 1 (PartialDerivative(p(mu)*q(nu), tau)._expand_partial_derivative() - PartialDerivative(p(mu), tau)*q(nu) - PartialDerivative(q(nu), tau)*p(mu))

/home/joha2/Programming/git/sympy/sympy/tensor/tensor.py in __sub__(self, other)
   1891 
   1892     def __sub__(self, other):
-> 1893         return TensAdd(self, -other).doit()
   1894 
   1895     def __rsub__(self, other):

/home/joha2/Programming/git/sympy/sympy/tensor/tensor.py in doit(self, **kwargs)
   2374                 return t.components, x.free, x.dum
   2375             return [], [], []
-> 2376         args.sort(key=sort_key)
   2377 
   2378         if not args:

/home/joha2/Programming/git/sympy/sympy/tensor/tensor.py in sort_key(t)
   2372             if hasattr(t, "_index_structure"):
   2373                 x = get_index_structure(t)
-> 2374                 return t.components, x.free, x.dum
   2375             return [], [], []
   2376         args.sort(key=sort_key)

/home/joha2/Programming/git/sympy/sympy/tensor/tensor.py in components(self)
   3310     @property
   3311     def components(self):
-> 3312         return self._get_components_from_args(self.args)
   3313 
   3314     @property

/home/joha2/Programming/git/sympy/sympy/tensor/tensor.py in _get_components_from_args(args)
   3209             if isinstance(arg, TensAdd):
   3210                 continue
-> 3211             components.extend(arg.components)
   3212         return components
   3213 

AttributeError: 'PartialDerivative' object has no attribute 'components'
```
Do you know where this could come from?

Upabjojr commented:
> Do you know where this could come from?

Yes, I do. The code is a mess. I will fix this.

joha2 commented:
:-) hehe. In the meanwhile I will prepare some unit tests to verify both functions working correctly. In a first step I will implement them with XFAIL. In a second step we can just remove it.

joha2 committed:
tensor: PartialDerivative: added _expand_partial_derivative() to TensExpr to delegate the expansion into sub expressions

joha2 committed:
tensor: PartialDerivative: constant factor rule and unit tests for _expand_partial_derivative()

joha2 committed:
tensor: PartialDerivative: try to fix tests

joha2 committed:
Merge branch 'master' of https://github.com/joha2/sympy into 18048_tensor_derivatives

oscarbenjamin commented:
The tests will not pass until #18238 is merged

joha2 commented:
@oscarbenjamin thanks for the hint!

joha2 committed:
Merge branch 'master' of https://github.com/sympy/sympy into 18048_tensor_derivatives

joha2 committed:
tensor: PartialDerivative: merged master, made tests for sum rule XFAIL due to failures in Python2.7

In Python2.7 certain simple tests fail for unknown reasons:

PartialDerivative(A(i) + B(i), tau).expand_partial_derivative() ==
PartialDerivative(A(i), tau) + PartialDerivative(B(i), tau) => FAIL

The tests do not fail in Python3.7

Upabjojr commented:
https://github.com/sympy/sympy/pull/18242

joha2 commented:
@Upabjojr thanks for the hint. Do you understand the errors arising on my last commit? It seems that occasionally some operations which work fine on Python3.7 do not work on other setups. (Notice, I had to XFAIL the check for the sum rule, and now the "constant coefficient" rule fails on pypy)

Upabjojr commented:
For some reason github is showing diffs from unrelated PR. Maybe we should switch to a new branch and open another PR.

joha2 committed:
Merge branch 'master' of https://github.com/sympy/sympy into 18048_tensor_derivatives

joha2 commented:
@Upabjojr Now that the tests are all working again, what about accepting the PR, and afterwards just adding tests in a new PR?

Upabjojr commented:
Actually the purpose of tests is to make sure the code being merged is correct.

SymPy requires every PR to be properly tested.

If we merge some code and then find out that it's not correct, that would be bad.

joha2 commented:
Yeah, understood. What is now the best way to proceed? Maybe the problems with the unrelated PRs comes from merging master quite frequently?

Upabjojr committed:
merge master

Upabjojr committed:
restore pretty

Upabjojr committed:
code cleanup

Upabjojr commented:
I've made some edits, I will look again in the next few days.

joha2 commented:
OK. But I did not get, what the changes actually do (except that they lead to failing tests :-)) Further, I don't know whether I am supposed to still add something to PR. Or do you intend to perform further changes in the next few days where my intervention would bring no improvement?

Upabjojr commented on sympy/tensor/tests/test_tensor_operators.py:
```diff
@@ -117,3 +123,183 @@ def test_replace_arrays_partial_derivative():
+    for count in range(10):
+        pos_random_int1 = sympify(randint(0, 1000))
+        pos_random_int2 = sympify(randint(0, 1000))
+        neg_random_int = sympify(randint(-1000, -1))
```
Maybe it's a bit exaggerated to test the same test 10 times... the random integers could be created only once.

joha2 replied:
I wanted to avoid that there are special cases appearing, which lead to a passing test one times and do not pass the next time. But yes, maybe 10 is exaggeration. :-)

Upabjojr replied:
Array operations can be tested like that, but I think it's not really necessary with symbolic expressions.

Upabjojr commented:
I have just cleaned up three code a bit. There were unnecessary changes plus the code was too complex. The type check should be done in the constructor of PartialDerivative, for example. 

Concerning the tests, I don't see the point in adding failing tests. Better fix them now.

Upabjojr committed:
Remove sorting in TensAdd printer. Remove for loop in test_tensor_operator.py tests

Upabjojr commented:
OK, there was a bug in `TensAdd._print` which was sorting the arguments before printing. The arguments used to appear in a different order.

I think this is almost ready for merging.

joha2 committed:
Merge branch 'master' of https://github.com/sympy/sympy into 18048_tensor_derivatives

joha2 commented:
I just updated the branch locally, and realized that the explicit derivatives are no more working. But they are still planned for this PR, aren't they?

Upabjojr commented:
> I just updated the branch locally, and realized that the explicit derivatives are no more working. But they are still planned for this PR, aren't they?

What do you mean? If something is not working it **must** be added to the tests with an assertion, and the code fixed.

joha2 commented:
@Upabjojr the `_eval_partial_derivative` is broken, after removing the type check, afaik. I will check that in more detail. But `_eval_partial_derivative` has to distinguish between `Symbol` and `Tensor`. What do you mean by implementing into the constructor? Could you please clarify on that?

Upabjojr commented:
> But `_eval_partial_derivative` has to distinguish between Symbol and Tensor.

Maybe `Symbol` should not be passed to `_eval_partial_derivative`. What about a separate method? 

> Or is the intention to implement the type check into PartialDerivative?

`PartialDerivative.__new__` is supposed to check the types. The private methods `_eval_<...>` should not check the types... they should be optimized for speed.

Upabjojr commented:
I think in the end `.doit()` and `.expand()` will call the private methods. Type checking should be done inside `.doit()`. I'm still not sure whether `.doit()` should have an optional parameter to switch Kronecker deltas on and off. What do you think?

joha2 commented:
> Maybe Symbol should not be passed to _eval_partial_derivative. What about a separate method?

The problem is that we introduce a lot of code doubling (as seen in _expand and _eval_partial_derivative right now) if we take care of tensors or symbols in different methods. An insitu type checking was a (not very elegant) solution for this. At the end of the day I want both: an abstract derivative like the one from expand and an explicit derivative for tensors and symbols for actual calculations.

> `PartialDerivative.__new__` is supposed to check the types. The private methods `_eval_<...>` should not check the types... they should be optimized for speed.

OK, that is a design decision, I can optimize my code for. So at the end of the day all eval methods should not have any type checkings?

> I'm still not sure whether .doit() should have an optional parameter to switch Kronecker deltas on and off. What do you think?

I am not a great fan of .doit() since for me it is still not clear, what it is supposed to do. Is it the counter part for calculations on hold (like (x + x).doit() = 2*x?)? Which kind of calculations have to be "doit()"? I have the feeling that it is sometimes performed implicitly, which is not good, if the user wants to have a certain calculation not to be performed. (Cf. our problems with the tests TensAdd(A(i), B(i)) vs. A(i) + B(i)). Let us assume that `doit()` for `PartialDerivative` has to be called explicitly (I am not sure about this), then a switch which type of `doit()` should be done (i.e. explicit for the Kroneckers or abstract for just the expansion, or any other type) would be very useful for sure.

So for reducing the amount of overhead for `_eval_partial_derivative`, how about this solution?
* first perform _expand() [which should stay an extra function]
* then in `__new__` it is checked for tensor or symbol and `diff` or `tensordiff` (`_eval_partial_derivative` in `Tensor`) is performed

The disadvantage is: The expansion can lead to many terms which for an explicit calculation are actually zero. What do you think about this solution?

Upabjojr commented:
> The problem is that we introduce a lot of code doubling

OK, so let's restore the type check. Remember to add a test to assure that this feature will work (some future contributors may otherwise inadvertently break the code).

joha2 committed:
tensor: PartialDerivative: reintroduced type checks for explicit derivatives

TODO: tests of explicit derivatives

Upabjojr commented:
Can you add a test for the case of derivation by scalars?

Upabjojr commented on sympy/tensor/tensor.py:
```diff
@@ -2624,6 +2638,8 @@ class Tensor(TensExpr):
     """
 
+    _diff_wrt = True
+    is_scalar = True
```
Are these two necessary?

joha2 replied:
Yes, they are, because if they are not added to tensor, the code complains that it does not know about how to derive tensors.

Upabjojr replied:
That means you are passing somewhere through `Derivative`. I would recommend not to allow this, it's correct that the code complains, because you should use `PartialDerivative` instead of `Derivative`.

joha2 replied:
Yeah, good point. It makes totally sense, to decouple these two. Within `_eval_partial_derivative` if there occurs a `Symbol` there is also `diff` invoked. Maybe there is some error occuring.

Upabjojr replied:
Don't call `diff`.

Upabjojr replied:
A tensor is not a scalar, this line is definitely wrong.

joha2 replied:
Is there any possibility to perform the derivative by symbols without using `diff`?

joha2 replied:
Yeah, I mentioned that at the beginning. Btw. a tensor of zeroth rank is scalar :-)

joha2 replied:
Ok removed

Upabjojr commented on sympy/tensor/tensor.py:
```diff
@@ -2997,6 +3013,46 @@ def _eval_rewrite_as_Indexed(self, tens, indices):
+
+            kronecker_delta_list = []
+            for iself, iother in zip(self.free_indices, s.free_indices):
+                if iself.tensor_index_type is not iother.tensor_index_type:
```
`!=` should be used.

`is not` checks if the objects occupy the same position in the computer memory... that is not a mathematical equality.

joha2 replied:
Oh sorry, I will change that.

joha2 commented:
I noticed another thing:
```Python
(PartialDerivative(q(nu)*p(mu), tau) - 
 PartialDerivative(p(mu), tau)*q(nu) - 
 PartialDerivative(q(nu), tau)*p(mu))._expand_partial_derivative().expand()
```
does not evaluate to zero, because it does not know that the tensors commute. How to implement this?

Upabjojr commented:
Tensors do not commute in general. You have to manually specify which tensors commute. `TensMul` handles that already.

There are special cases, like spinors, which anti-commute. So you have `A*B == -B*A` for spinors. Some other objects have no commutation relations at all.

joha2 commented:
OK, I will have a look. Thanks! I'm implementing the tests right now: Do you know why this assertion failes?
```Python
expr1 = PartialDerivartive(A(i), A(j))
assert expr1._eval_partial_derivative() == L.delta(i, -j)
```
Somehow sympy does not get that `L.delta(i, -j)` equals `KD(i, -j)`.

joha2 committed:
tensor: PartialDerivative: added tests for _eval_partial_derivative

Further, restored index check in explicit tensor derivative.

Upabjojr commented:
> Therefore the Kronecker deltas do not get the right indices. It seems that `free_indices` return a set, but I need a list with the correct order.
> 
> Edit: Used `self.args[1]` since here the order is preserved. For the API, I would suggest that `free_indices` are also a tuple with preserved order.

Have you tried using `.get_free_indices()`? The current tensor API needs to be reviewed a bit... I recommend `.get_free_indices()` and `.get_indices()`.

joha2 committed:
tensor: PartialDerivative: checks for unused symbols fixed

joha2 commented:
OK, I will check that. Thanks for the hint!

joha2 committed:
tensor: PartialDerivative: substituted args[1] by get_free_indices() in _eval_partial_derivatives()

joha2 committed:
tensor: PartialDerivative: removed _diff_wrt = True, is_scalar = True, took care of edge cases in _eval_partial_derivative

joha2 commented:
I don't understand, why these tests are sometimes failing and sometimes not. Are we sure, that the order of the summands is reproducible? Or do we have to call some expand before they can be compared?

Upabjojr commented on sympy/tensor/tensor.py:
```diff
@@ -2598,12 +2598,16 @@ def _eval_rewrite_as_Indexed(self, *args):
+            else:
+                # do not call diff if s is no symbol
+                if isinstance(s, Symbol):
+                    my_diffs.append(a.diff(s))
```
`a._eval_derivative(s)`... `a.diff(s)` is quite expensive.

joha2 replied:
Good to know. That's my missing knowledge of the underlying API :-) Thanks for fixing.

Upabjojr committed:
various fixes

Upabjojr committed:
fixing tests by sorting args in TensAdd

Upabjojr committed:
fixing test_pretty

joha2 commented:
Thanks for fixing this! :smile: 
Is it ready for merge? :grimacing:

Upabjojr commented:
I think so. Did you add an assert to every intended use case in the tests?

It's really important to do so because someone else in the future may edit the code and disrupt some existing functionality without realizing it. Unit tests are intended to prevent it.

joha2 commented:
For comparison, we can check whether some functionality is tested, which is not, and which is not necessary:

Standard rules:

* Linearity (OK, including constant factors, including second derivatives)
* Product rule (OK, including second derivatives)
* Consistency: Comparison between different approaches of performing (see tensor_diff.pdf) (not yet)

Higher order derivatives:

* Flattening (as already tested)
* Symmetry (**is there a canonicalization of PartialDerivative of higher order?**)

Contractions:

* Inner (divergence type) (OK for first order, still some issue open, see next comment)
* Outer (vector gradient type) (OK, components)
* Vector analysis identities (div rot = 0, rot grad = 0, **works only if derivatives are commuting**)
* Contractions with Levi-Civita-tensor (**is Levi-Civita-tensor already defined?**)

joha2 commented:
@Upabjojr I checked the tests and realized that there is one line which seems to be not correct:
```Python
    expr4 = PartialDerivative(H(i, j), H(-m, -m1))
    print(expr4._perform_derivative())
```
leads to
```
metric(i, L_0)*KD(-L_0, m)*metric(j, L_0)*KD(-L_0, m1)
```
where it seems that `L_0` is occuring multiple times, which may lead to inconsistent expression after canonicalization, metric and delta contraction. Do you agree?

It probably comes from the generation of the dummy index in `Tensor._eval_partial_derivative`, where for every inserted metric a dummy index between metric and Kronecker delta is needed. Is it possible to generate dummy indices with different names on the fly?

jksuom commented:
> is Levi-Civita-tensor already defined?

There is an implementation in sympy/functions/special/tensor_functions.py.

Upabjojr commented:
> It probably comes from the generation of the dummy index in `Tensor._eval_partial_derivative`, where for every inserted metric a dummy index between metric and Kronecker delta is needed. Is it possible to generate dummy indices with different names on the fly?

The algorithm is already supposed to generate dummy indices with different names. I'm not sure what's happening there.

joha2 commented:
It is the old problem of using `*` instead of `TensMul`
```Python
                        kroneckerdelta = (
                            TensMul(tensor_metric(iself, dummy),
                                    tensor_index_type.delta(-dummy, -iother))
                        )
```
leads to the correct (self generated) dummy index structure
```
metric(i, d_0)*KD(-d_0, m)*metric(j, d_1)*KD(-d_1, m1)
```
where using `*` leads to the inconsistent structure. But I am not sure, whether ``L_0`` in the first metric and ``L_0`` in the second metric are really the same. At least it is shown like that and the test also does not fail.

joha2 committed:
tensor: PartialDerivative: fixed inconsistent index generation in tensor derivative

The multiplication `*` seems to substitute all summation indices in an expression to dummies, which
is of no use if an expression is formed by generating each factor individually. Therefore, TensMul
was invoked directly, such that the expression does not get its dummy indices renamed before finalization.

joha2 commented:
@Upabjojr I proposed a fix, please review it. From an API perspective it should not even be possible to generate an expression like `A(L_0)*B(-L_0)*C(L_0)*D(-L_0)`, which I accidentially did in the tests. Can there a test be added to determine, whether an index is only used in exact two positions?

joha2 commented:
@jksuom thanks for the hint! I will have a look. I realized that there is also epsilon property in the `TensorIndexType` class.

joha2 committed:
tensor: PartialDerivative: fixed another index error in tensor derivative calculation, added doit()

doit() is necessary to rename double indices in proper numbered dummys. Added some tests for
divergence type derivatives including consistency checks.

Upabjojr commented:
I've merged this PR as it is too long. 

If there are other issues, you can open another PR.

Thanks for contributing!

joha2 commented:
@Upabjojr thanks for merging! :smile: Now we have the base functionality. There are still some items left on my wishlist. I would have been opened new issues/PRs for them anyway. Thank you for your thorough support!
